### PR TITLE
delta sharing server changes for json predicates.

### PR DESCRIPTION
**Pull request by chakankardb**
This PR implements the json predicate based filtering in the delta sharing server.
The changes are disabled by default.

### CODE REVIEW

File: server/src/main/scala/io/delta/standalone/internal/JsonPredicates.scala [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+      case OpDataTypes.IntType => leftVal.toInt < rightVal.toInt
+      case OpDataTypes.LongType => leftVal.toLong < rightVal.toLong
+      case OpDataTypes.StringType => leftVal < rightVal
+      case OpDataTypes.DateType =>
+        java.sql.Date.valueOf(leftVal).before(java.sql.Date.valueOf(rightVal))
+      case _ =>
+        throw new IllegalArgumentException("Unsupported type: " + leftType)
+    }
+  }
+}
```
Comment (chakankardb) [inline server/src/main/scala/io/delta/standalone/internal/JsonPredicates.scala:417]: This file is mostly copied from spark connector.

File: server/src/test/scala/io/delta/standalone/internal/JsonPredicateSuite.scala [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+    val numEval = 1000000
+
+    val start = System.currentTimeMillis()
+    for (i <- 1 to numEval) {
+      evalExpectBoolean(op, EvalContext(Map("date" -> dates(i % dates.size))))
+    }
+    val t = System.currentTimeMillis() - start
+    println("Stress test took " + t + " millis")
+  }
+}
```
Comment (chakankardb) [inline server/src/test/scala/io/delta/standalone/internal/JsonPredicateSuite.scala:471]: copied from spark connector.

**Review by wchau — APPROVED** [review]

**Comment by chakankardb** [comment]
@linzhou-db You want to take a quick look?

File: server/src/main/scala/io/delta/standalone/internal/JsonPredicates.scala [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+ * The evaluation context in which operations will be evaluated.
+ *
+ * @param partitionValues The partition values associated with parquet files.
+ */
+case class EvalContext(partitionValues: Map[String, String])
+
+/**
+ * The data types supported by the filtering operations.
+ */
+object OpDataTypes {
```
Comment (linzhou-db) [inline server/src/main/scala/io/delta/standalone/internal/JsonPredicates.scala:36]: Seems like there's no way to use the same Jsonpredicates in both server/ and spark/?

**Review by linzhou-db — APPROVED** [review]

File: server/src/main/scala/io/delta/standalone/internal/JsonPredicates.scala (hunk 3 above)
Comment (chakankardb) [reply server/src/main/scala/io/delta/standalone/internal/JsonPredicates.scala:36]: Ya, I had checked with Ryan Zu as well.
The way would be to make a common libraries, but that will trigger additional dependencies.
We can do that as a follow up change.

**Pull request merged**